tools/autograd/templates/Functions.cpp

```diff
@@ -1792,6 +1792,55 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
   return u_term + sigma_term + v_term;
 }
 
+// "An extended collection of matrix derivative results for forward and reverse mode algorithmic differentiation"
+// https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
+Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
+                    bool eigenvectors, const Tensor& lambda, const Tensor& v) {
+  // This gradient only works for real eigenvalues at the moment.
+  TORCH_CHECK(eigenvectors,
+           "eig_backward: Setting eigenvectors to false in torch.eig doesn't compute eigenvectors ",
+           "and hence we cannot compute backward. Please use torch.eig(eigenvectors=True)");
+  auto zeros = at::zeros({1}, lambda.options());
+  TORCH_CHECK(
+      at::allclose(lambda.slice(/*dim=*/-1, /*start=*/1, /*end=*/2), zeros),
+      "eig_backward: Backward calculation does not support complex eigenvalues at the moment.");
+
+  auto glambda = grads[0];
+  auto gv = grads[1];
+  auto vt = v.transpose(-2, -1);
+
+  Tensor result;
+  // contribution from the eigenvectors
+  if (gv.defined()) {
+    auto rlambda = lambda.slice(/*dim=*/-1, /*start=*/0, /*end=*/1);
+
+    auto hm = rlambda.transpose(-2,-1) - rlambda;
+    hm.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(INFINITY);
+    hm.pow_(-1.0);
+
+    auto gvortho = gv - at::sum(gv * v, /*dim=*/-2, /*keepdim=*/true) * v;
+    auto B = hm * at::matmul(vt, gvortho);
```

Comment:
@mfkasim1 , could you please explain this line? It does not seem to agree with the reference you mentioned above.

Reply:
I actually derived the expression myself, but I didn't have the time to write it down, so I just put the closest available literature in the code.

Reply:
The derivations are different, so I do not know which method to trust, but I do not immediately see any issues in the reference. So, then, how did you get that `gvortho`?

Reply:
I'm not sure if they are actually different or actually the same with just different expressions.
I have put my expression into `gradcheck` and `gradgradcheck` and it passes.
I will write the derivation down if I got time.

Reply:
They are different, in your derivation you subtract from the `gv` the projection of `gv` into `v`, and they do not do it in the reference.

Reply:
Ah, if I remember it correctly, I subtract it because of the constraint of the eigenvectors to have `norm = 1`.
I probably did the similar trick from [this paper](https://arxiv.org/abs/2011.04366) (eq. 3.23-3.25)

Reply:
All right, now I see. I did check the output of `torch.eig` it indeed produces normalized eigenvectors. Awesome, thank you!